### PR TITLE
add new stabilization parameter for entropy viscosity

### PR DESCRIPTION
**Pull request by jdannberg**
The entropy viscosity stabilization previously did not work very well for models that were dominated by a shear strain rate that was applied externally (as opposed to velocities arising from convection), see #967.
This pull request fixes this by adding a parameter that allows the entropy viscosity to also be scaled with the strain rate * cell diameter in addition to the maximum velocity in a cell. The default value for this parameter is zero, which means that in this case the entropy viscosity computation should be the same as before this change.

### CODE REVIEW

**Comment by jdannberg** [comment]
The attached screenshot shows the difference between the old (top) and the new (bottom) version for a shear bands model.
stabilization

**Comment by tjhei** [comment]
Very nice!

1. Do we have a section in the manual that defines the EV method? If not we should. If yes, can you add this parameter?
2. What default for gamma would you pick if you don't care about the tests? Do we have an intuition if that hurts or helps in general?
3. We should create an issue to generate a suite of tests for the stabilization (sinker, etc.) to see what a good default is. Agreed?

File: doc/modules/changes.h [hunk 1]
```diff
@@ -6,6 +6,12 @@
  *
  * <ol>
  *
+ * <li> New: There is a new parameter in the entropy viscosity stabilization
+ * that allows to scale the stabilization with the strain rate (in addition
+ * to the velocity).
```
Comment (bangerth) [inline doc/modules/changes.h]: might be worth to mention it by name

File: source/simulator/assembly.cc [hunk 2]
```diff
@@ -578,14 +578,20 @@ namespace aspect
         const Tensor<1,dim> velocity = (scratch.old_velocity_values[q] +
                                         scratch.old_old_velocity_values[q]) / 2;
 
+        const SymmetricTensor<2,dim> strain_rate_tensor = (scratch.old_strain_rates[q] +
+                                                           scratch.old_old_strain_rates[q]) / 2;
+        const double strain_rate = std::sqrt(strain_rate_tensor * strain_rate_tensor);
```
Comment (bangerth) [inline source/simulator/assembly.cc]: I think this is `strain_rate_tensor.norm()`. You may not even need to name the strain rate tensor object if you just write
```
  const double ... = ((old_s_r[q] + o_o_strain_r[q])/2).norm();
```

File: source/simulator/assembly.cc [hunk 3]
```diff
@@ -578,14 +578,20 @@ namespace aspect
...
+
         if (parameters.stabilization_alpha == 2)
           {
             const double field = (scratch.old_field_values[q] + scratch.old_old_field_values[q]) / 2;
             residual[q] *= std::abs(field - average_field);
           }
 
         max_residual = std::max (residual[q],     max_residual);
-        max_velocity = std::max (std::sqrt (velocity*velocity), max_velocity);
+        max_velocity = std::max (std::sqrt (velocity*velocity)
```
Comment (bangerth) [inline source/simulator/assembly.cc]: while there, you could also use `velocity.norm()` here

File: source/simulator/parameters.cc [hunk 4]
```diff
@@ -664,6 +664,15 @@ namespace aspect
                            "different value than described there: It can be chosen as stated there for "
                            "uniformly refined meshes, but it needs to be chosen larger if the mesh has "
                            "cells that are not squares or cubes.) Units: None.");
+        prm.declare_entry ("gamma", "0.0",
+                           Patterns::Double (0),
+                           "The strain rate scaling factor in the artificial viscosity "
+                           "stabilization. This parameter determines how much the strain rate (in addition "
+                           "to the velocity) should influence the stabilization. (This parameter does "
+                           "not correspond to any variable in the 2012 GJI paper by Kronbichler, "
+                           "Heister and Bangerth that describes ASPECT. Rather, the paper always uses "
```
Comment (bangerth) [inline source/simulator/parameters.cc]: I think it would be good to spell out the formula, corresponding to the `\nu^\text{max}` thing in the paper.

**Comment by bangerth** [comment]
OK with the modifications mentioned, and once the test actually passes the tester.

**Comment by jdannberg** [comment]
To reply to @tjhei:
1/ We do not have a section in the manual (only one about smoothing of entropy viscosity). And yes, we should add one. See #1039.
2/ I have no intuition for how to pick gamma in general. This would be something to look at in a suite of tests for the stabilization, so...
3/ Yes, definitely. See #1040.

**Comment by bangerth** [comment]
On 2/ Intuitively, a factor gamma=1 will lead to an additional term that is roughly the same size as `||u||_\infty` on cells where the velocity vanished at one point. On cells where `u` is not small the additional term may not make much of a difference. I suspect that choosing gamma between 0 and 1 would be a good choice.

**Comment by bangerth** [comment]
What do you choose it like in the benchmark?

**Comment by jdannberg** [comment]
I tried something like 1 first (because that was also what I thought I should choose from how we set this value for the models we ran before), but it turned out to be much too small. The model I showed to you was with a value of 10. But we never ran the benchmark case with such a low resolution, so for a reasonable resolution maybe something around 1 would be fine.

**Comment by yinghe616** [comment]
Just want to update that I run the sinker-average problem with AMR turned on tonight, and with either gamma=1 or =2, the results doesn't change too much which gives an overshoot/undershoot of 12.5% on a level 10 refinement with refine/coarsen fraction number (0.07, 0.01). I will show the numerical results during my presentation tomorrow. 
Also I would be interested to see if any changes on your new setting.

**Comment by bangerth** [comment]
I'd say ten is approximately equal to 1 :-) I also wouldn't be surprised if you could get away with a value of 4 or so. Either way, in places where the velocity is large, it won't make a difference.

I'm still surprised that the patch that the Clemson people implemented and that takes the max over all neighboring cells doesn't work. It would correspond to a value of gamma=2.

**Pull request merged**